src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -0,0 +1,644 @@
+//       arr[i + 5] = w; // bounds check
+//       arr[i + 8] = w; // bounds check
+//   }
+//
+// The phase scans all statements in a block and groups the bounds checks based on
+// "Base Index and Length" pairs (VNs). Then the phase takes the largest group and
+// clones the block to have fast and slow paths.
+
+//------------------------------------------------------------------------------------
+// Initialize: Initialize the BoundsCheckInfo with the given bounds check node.
+//    and perform some basic legality checks.
+//
+// Arguments:
+//    comp             - The compiler instance
+//    statement        - The statement containing the bounds check
+//    bndChkNode       - The bounds check node
+//    bndChkParentNode - The parent node of the bounds check node (either null or COMMA)
+//
+// Return Value:
+//    true if the initialization was successful, false otherwise.
+//
+bool BoundsCheckInfo::Initialize(const Compiler*   comp,
+                                 Statement*        statement,
+                                 GenTreeBoundsChk* bndChkNode,
+                                 GenTree*          bndChkParentNode)
+{
+    idxVN = comp->vnStore->VNConservativeNormalValue(bndChkNode->GetIndex()->gtVNPair);
+    lenVN = comp->vnStore->VNConservativeNormalValue(bndChkNode->GetArrayLength()->gtVNPair);
+    if ((idxVN == ValueNumStore::NoVN) || (lenVN == ValueNumStore::NoVN))
+    {
+        return false;
+    }
+
+    stmt         = statement;
+    bndChk       = bndChkNode;
+    bndChkParent = bndChkParentNode;
+
+    assert((bndChkParent == nullptr) || bndChkParent->OperIs(GT_COMMA));
+
+    if (bndChkNode->GetIndex()->IsIntCnsFitsInI32())
+    {
+        // Index being a constant means we have index=0 and cns offset
+        offset = static_cast<int>(bndChkNode->GetIndex()->AsIntCon()->IconValue());
+        idxVN  = comp->vnStore->VNZeroForType(TYP_INT);
+    }
+    else
+    {
+        if (comp->vnStore->TypeOfVN(idxVN) != TYP_INT)
+        {
+            return false;
+        }
+
+        // Otherwise, peel the offset from the index using VN
+        comp->vnStore->PeelOffsetsI32(&idxVN, &offset);
+        assert(idxVN != ValueNumStore::NoVN);
+    }
+    assert(comp->vnStore->TypeOfVN(idxVN) == TYP_INT);
+
+    if (offset < 0)
+    {
+        // Not supported yet.
+        return false;
+    }
+    return true;
+}
+
+//------------------------------------------------------------------------------------
+// RemoveBoundsChk - Remove the given bounds check from the statement and the block.
+//    This function is a copy of optRemoveRangeCheck, but without questionable
+//    relaxations around side effect extractions.
+//
+// Arguments:
+//    comp  - compiler instance
+//    check - bounds check node to remove
+//    comma - check's parent node (either null or COMMA)
+//    stmt  - statement containing the bounds check
+//
+static void RemoveBoundsChk(Compiler* comp, GenTreeBoundsChk* check, GenTree* comma, Statement* stmt)
```

Comment:
If you are copying this out anyway then why not change it to support any bounds check, by just taking the `GenTree**` for the bounds check?

Reply:
@jakobbotsch can you elaborate? what kind of benchmarks it does not support now?

Reply:
Any embedded statement that isn't `GT_COMMA` (e.g. `GT_CALL` setup nodes).
I think it would be an overall simplification of the code to stop caring about the parent of the bounds check.

Reply:
> Any embedded statement that isn't GT_COMMA (e.g. GT_CALL setup nodes).

sorry, still don't understand. We have a bounds check that can be either root (then its user is Statement, not some other node) or COMMA (in which case we can re-use the comma to propagate side effects from the bounds check) - how can `GenTree**` argument help here?

Reply:
If a `BOUNDS_CHECK` ever ends up as the setup node in `GT_CALL` then your pass will hit this assertion:
https://github.com/dotnet/runtime/blob/8f867f64556bf267d68376230134642a84a8e4c6/src/coreclr/jit/rangecheckcloning.cpp#L73
It should be legal IR to have any embedded statement as a setup node in calls. If we ever end up with other forms of embedded statements, then it should also be legal to have bounds checks inside those.

To fix this, I think you can just remove `BoundCheckLocation::bndChkParent` and change `BoundCheckLocation::bndChk` to a `GenTree**` that points to the use (which you save in `BoundsCheckVisitor::PostOrderVisit`). Then you have the edge already on hand when you need to remove the bounds checks in this function.

Reply:
Ah, thanks! I thought about it too, but I assumed that I would need the actual parent anyway to use it as a split point for the last check, but I presume I can just find it by hands if I ever need it

Reply:
Thanks for the suggestion - addressed!